contentcuration/contentcuration/middleware/StudioLocaleMiddleware.py

```diff
@@ -0,0 +1,160 @@
+from django.utils.translation.trans_real import check_for_language
+from django.utils.translation.trans_real import get_language_from_path
+from django.utils.translation.trans_real import get_languages
+from django.utils.translation.trans_real import get_supported_language_variant
+from django.utils.translation.trans_real import language_code_re
+from django.utils.translation.trans_real import parse_accept_lang_header
+
+
+def get_accept_headers_language(request):
+    accept = request.META.get("HTTP_ACCEPT_LANGUAGE", "")
+    for accept_lang, unused in parse_accept_lang_header(accept):
+        if accept_lang == "*":
+            break
+
+        if not language_code_re.search(accept_lang):
+            continue
+
+        try:
+            return get_supported_language_variant(accept_lang)
+        except LookupError:
+            continue
+
+
+def get_language_from_request_and_is_from_path(request):  # noqa complexity-16
+    """
+    Analyzes the request to find what language the user wants the system to
+    show. Only languages listed in settings.LANGUAGES are taken into account.
+    If the user requests a sublanguage where we have a main language, we send
+    out the main language. It also returns a value to determine if the language code
+    was derived from a language code in the URL, or inferred from some other source.
+    :returns: tuple of language code, boolean. The former can be None if the url being
+    requested does not require translation, otherwise it should be a language code
+    from the values in settings.LANGUAGES. The boolean should indicate whether the
+    language code was calculated by reading a language code from the requested URL.
+    In the case that it was, True should be returned, in the case where the URL language
+    code was not used or not present, False is returned.
+    """
+
+    try:
+        # If this is not a view that needs to be translated, return None, and be done with it!
+        if not getattr(resolve(request.path_info).func, "translated", False):
+            return None, False
+    except Resolver404:
+        # If this is an unrecognized URL, it may be redirectable to a language prefixed
+        # URL, so let the language code setting carry on from here.
+        pass
+
+    supported_lang_codes = get_languages()
+
+    lang_code = get_language_from_path(request.path_info)
+    if lang_code in supported_lang_codes and lang_code is not None:
+        return lang_code, True
+
+    if hasattr(request, "session"):
+        lang_code = request.session.get(LANGUAGE_SESSION_KEY)
+        if (
+            lang_code in supported_lang_codes
+            and lang_code is not None
+            and check_for_language(lang_code)
+        ):
+            return lang_code, False
+
+    headers_language = get_accept_headers_language(request)
+
+    if headers_language is not None:
+        return headers_language, False
+
+    return get_settings_language(), False
```

Comment:
is this something we also use in Kolibri? if so, maybe should go in our kolibri utils Python package?

Reply:
Right, that's what I meant. kolibri-utils, kolibri-tools, le-utils... basic idea would be to have some place for shared code